File: finder4j-backtrack-examples/src/main/java/com/github/K0zka/finder4j/backtrack/examples/lab/LabStep.java
package com.github.K0zka.finder4j.backtrack.examples.lab;

import java.util.ArrayList;

import com.github.K0zka.finder4j.backtrack.Step;

@SuppressWarnings("serial")
public class LabStep implements Step<LabState> {

	public LabStep(Position stepTo) {
		super();
		this.stepTo = stepTo;
	}

	private final Position stepTo;

	public LabState take(LabState state) {
		final ArrayList<Position> newRoute = new ArrayList<Position>(
				state.getRoute());
		newRoute.add(stepTo);
		return new LabState(state.getLab(), newRoute);
	}

	@Override
	public String toString() {
		return "{ -> " + stepTo + "}";
	}

	Position getStepTo() {
		return stepTo;
	}

}

File: finder4j-backtrack-examples/src/main/kotlin/com/github/k0zka/finder4j/backtrack/examples/lab/LabStep.kt
package com.github.k0zka.finder4j.backtrack.examples.lab

import com.github.k0zka.finder4j.backtrack.Step
import java.util.ArrayList

class LabStep(internal val stepTo: Position) : Step<LabState> {

	override fun take(state: LabState): LabState {
		val newRoute = ArrayList(
				state.route)
		newRoute.add(stepTo)
		return LabState(state.lab, newRoute)
	}

	override fun toString(): String {
		return "{ -> $stepTo}"
	}

}
